### PR TITLE
🐛 fix missing asset name during local scan reporting upstream

### PR DESCRIPTION
**Pull request by vjeffrey**
fixes https://github.com/mondoohq/server/issues/6494

### CODE REVIEW

File: providers/os/provider/provider.go [hunk 1]
```diff
@@ -325,6 +325,9 @@ func (s *Service) connect(req *plugin.ConnectReq, callback plugin.ProviderCallba
 			asset.Name = fingerprint.Name
 			asset.PlatformIds = fingerprint.PlatformIDs
 		}
+		if asset.Name == "" && len(asset.Connections) > 0 {
+			asset.Name = asset.Connections[0].Host
```
Comment (czunker) [inline providers/os/provider/provider.go]: When checking this in debug on my local system, `asset.Connections[0].Host` is empty.

**Comment by czunker** [comment]
I'm not sure this is a cnquery bug. When I run cnquery locally on an AWS EC2 RHEL 8 instance, I get this:
```
cnquery run -c "asset{ name ids }"
→ installing provider 'os' version=9.0.12
→ successfully installed os provider path=/home/ec2-user/.config/mondoo/providers/os version=9.0.12
→ no provider specified, defaulting to local. Use --help to see all providers.
→ no Mondoo configuration file provided, using defaults
asset: {
  ids: [
    0: "//platformid.api.mondoo.app/hostname/ip-172-31-21-4.eu-central-1.compute.internal"
    1: "//platformid.api.mondoo.app/runtime/aws/ec2/v1/accounts/735798807192/regions/eu-central-1/instances/i-0fd993fead5d6b1d1"
  ]
  name: "i-0fd993fead5d6b1d1"
}
```
But when I run cnspec on the same machine:
```
cnspec scan local --config ./sa-edge.json 
→ loaded configuration from ./sa-edge.json using source --config
→ using service account credentials
→ discover related assets for 1 asset(s)
→ synchronize assets

  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% score


Asset: //assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq
---------------------------------------------------------------------------------------------

Data queries:
....
asset: {
  kind: "baremetal"
  title: "Red Hat Enterprise Linux 8.4 (Ootpa), bare metal"
  arch: "x86_64"
  platform: "redhat"
  runtime: ""
  name: "i-0fd993fead5d6b1d1"
}
....
Scanned 1 assets

Red Hat Enterprise Linux 8.4 (Ootpa)
    C 

```

The data queries determine the correct asset name, but the progress bar, the "Asset XYZ" title, and the summary are missing the asset name.

**Comment by vjeffrey** [comment]
hmm, interesting. when i tested this bug i spun up a rhel, sshd to it, and then ran a local scan, reporting upstream. i did see the asset name missing as reported. when i added this fix the asset name reappeared

**Comment by czunker** [comment]
> hmm, interesting. when i tested this bug i spun up a rhel, sshd to it, and then ran a local scan, reporting upstream. i did see the asset name missing as reported. when i added this fix the asset name reappeared

I gave it another try, but still no success.
AMI: amazon/RHEL_HA-8.4.0_HVM-20230419-x86_64-41-Hourly2-GP2
I build the os provider from this branch locally and copied it to the machine.
I ran cnspec on the VM.
It used the copied provider:
```
cnspec providers list

→ builtin (found 2 providers)

  core 9.0.2
  mock 9.0.0 with connectors: mock

→ /home/ec2-user/.config/mondoo/providers (found 1 providers)

  os 9.0.12 with connectors: local, ssh, winrm, vagrant, container, docker, filesystem
```

**Comment by czunker** [comment]
Strange is, that cnspec has the other information:
```
→ no provider specified, defaulting to local. Use --help to see all providers.
DBG using provider os with connector local
DBG Started a new runtime (2 total)
DBG no need to update provider last-refresh=7m35.891518298s provider=os
DBG running provider plugin path=/home/ec2-user/.config/mondoo/providers/os/os
DBG Started a new runtime (3 total)
→ loaded configuration from ./sa-edge.json using source --config
→ using service account credentials
DBG Started a new runtime (4 total)
→ discover related assets for 1 asset(s)
DBG Started a new runtime (5 total)
DBG running provider plugin path=/home/ec2-user/.config/mondoo/providers/os/os
→ synchronize assets
DBG initialize client authentication issuer=mondoo/ams kid=//agents.api.mondoo.app/spaces/elastic-payne-456551/serviceaccounts/2VhazWr7giaxxTUg9yykztRVg70
DBG got assets details assets=5
DBG asset mapping asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq platform-mrn=//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-xyDppn0lX4XfcPi779pmzQsplKby7DP+krtkLk42xaM
DBG asset mapping asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq platform-mrn=//platformid.api.mondoo.app/hostname/ip-172-31-21-4.eu-central-1.compute.internal
DBG asset mapping asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq platform-mrn=//platformid.api.mondoo.app/runtime/aws/ec2/v1/accounts/735798807192/regions/eu-central-1/instances/i-0fd993fead5d6b1d1
DBG asset mapping asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq platform-mrn=//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-d++yp4tI30LK+kKy+qa+KU0iXvDgm3TBU1DWp+U7fYw
DBG asset mapping asset=//assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq platform-mrn=//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-ics+d+ynhnMcrqPc5vB2wh6TvAe48gsu4ZVI8R9Ssy0
DBG update asset asset= platform-ids=["//platformid.api.mondoo.app/hostname/ip-172-31-21-4.eu-central-1.compute.internal","//platformid.api.mondoo.app/runtime/aws/ec2/v1/accounts/735798807192/regions/eu-central-1/instances/i-0fd993fead5d6b1d1","//platformid.api.mondoo.app/runtime/aws/ec2/v1/accounts/735798807192/regions/eu-central-1/instances/i-0fd993fead5d6b1d1","//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-d++yp4tI30LK+kKy+qa+KU0iXvDgm3TBU1DWp+U7fYw","//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-ics+d+ynhnMcrqPc5vB2wh6TvAe48gsu4ZVI8R9Ssy0","//platformid.api.mondoo.app/runtime/ssh/hostkey/SHA256-xyDppn0lX4XfcPi779pmzQsplKby7DP+krtkLk42xaM"]
DBG connecting to asset  (Red Hat Enterprise Linux 8.4 (Ootpa))
```

And one of the ids also includes the instance id, which should be the asset name.

**Comment by czunker** [comment]
I got it working:
```
Red Hat Enterprise Linux 8.4 (Ootpa)
    C i-0fd993fead5d6b1d1
```
But only when using a builtin provider:
```
cnspec providers list
! can't find any paths for providers, none are configured home-path=/home/ec2-user/.config/mondoo/providers system-path=/opt/mondoo/providers

→ builtin (found 3 providers)

  core 9.0.2
  mock 9.0.0 with connectors: mock
  os 9.0.12 with connectors: local, ssh, winrm, vagrant, container, docker, filesystem

→ /home/ec2-user/.config/mondoo/providers has no providers

→ /opt/mondoo/providers has no providers
```

So, perhaps cnspec is using a different asset than cnquery and we are saving the name in the different asset?

**Comment by vjeffrey** [comment]
but, wouldn't we need to use builtin provider to test this, since its a change in the os provider?

**Comment by chris-rock** [comment]
you can always test without built in. Just run `make providers/build providers/install`

**Comment by czunker** [comment]
> but, wouldn't we need to use builtin provider to test this, since its a change in the os provider?

What I tried to say with the previous comments:
I only see the asset name in the output of cnspec, when I use the os provider builtin. When I use the os provider as plugin, I still do not see the asset name, even with the fix in this PR. That let's me think, that this PR isn't the actual fix. But I'm not really sure. I wonder why it is working on my local system. So, really strange. But I can also remember we issues in the last weeks, where things were working when we used providers builtin, but not then used them as plugin.

**Comment by vjeffrey** [comment]
oh, that is very odd..hmm

**Comment by czunker** [comment]
I could also reproduce this with cnquery only:
cnquery scan --config sa-edge.json 
! can't find any paths for providers, none are configured home-path=/home/ec2-user/.config/mondoo/providers system-path=/opt/mondoo/providers
→ installing provider 'os' version=9.0.12
→ successfully installed os provider path=/home/ec2-user/.config/mondoo/providers/os version=9.0.12
→ no provider specified, defaulting to local. Use --help to see all providers.
...
→ discover related assets for 1 asset(s)
→ synchronize assets

...
Asset: 
-------

...
Retrieve asset information:
asset: {
  kind: "baremetal"
  title: "Red Hat Enterprise Linux 8.4 (Ootpa), bare metal"
  arch: "x86_64"
  platform: "redhat"
  runtime: ""
  name: "i-0fd993fead5d6b1d1"
}
...

Retrieve the hostname:
os.hostname: "ip-172-31-21-4.eu-central-1.compute.internal"

Retrieve the platform:
asset.platform: "redhat"

...
Summary (1 assets)
==================

Target:     //assets.api.mondoo.app/spaces/elastic-payne-456551/assets/2WeKBIbfSFG20uYEs1FyBviWYKq
Datapoints: 24

```

**Comment by czunker** [comment]
With the latest commit, it now looks like this:
```
Summary (1 assets)
==================

Target:     i-0fd993fead5d6b1d1
Datapoints: 24
```

**Review by jaym — APPROVED** [review]

**Pull request merged**